common/archiver/historyIterator.go

```diff
@@ -0,0 +1,287 @@
+		runID:             request.RunID,
+		shardID:           request.ShardID,
+		eventStoreVersion: request.EventStoreVersion,
+		branchToken:       request.BranchToken,
+		config:            config,
+	}
+	if it.sizeEstimator == nil {
+		it.sizeEstimator = NewJSONSizeEstimator()
+	}
+	if initialState == nil {
+		return it, nil
+	}
+	if err := it.reset(initialState); err != nil {
+		return nil, err
+	}
+	return it, nil
+}
+
+func (i *historyIterator) Next() (*HistoryBlob, error) {
+	if !i.HasNext() {
+		return nil, errIteratorDepleted
+	}
+
+	historyBatches, newIterState, err := i.readHistoryBatches(i.NextEventID)
+	if err != nil {
+		return nil, err
+	}
+
+	i.historyIteratorState = newIterState
+	firstEvent := historyBatches[0].Events[0]
+	lastBatch := historyBatches[len(historyBatches)-1]
+	lastEvent := lastBatch.Events[len(lastBatch.Events)-1]
+	eventCount := int64(0)
+	for _, batch := range historyBatches {
+		eventCount += int64(len(batch.Events))
+	}
+	header := &HistoryBlobHeader{
+		DomainName:           common.StringPtr(i.domainName),
+		DomainID:             common.StringPtr(i.domainID),
+		WorkflowID:           common.StringPtr(i.workflowID),
+		RunID:                common.StringPtr(i.runID),
+		IsLast:               common.BoolPtr(i.FinishedIteration),
+		FirstFailoverVersion: firstEvent.Version,
+		LastFailoverVersion:  lastEvent.Version,
+		FirstEventID:         firstEvent.EventId,
+		LastEventID:          lastEvent.EventId,
+		EventCount:           common.Int64Ptr(eventCount),
+	}
+
+	return &HistoryBlob{
+		Header: header,
+		Body:   historyBatches,
+	}, nil
+}
+
+// HasNext returns true if there are more items to iterate over.
+func (i *historyIterator) HasNext() bool {
+	return !i.FinishedIteration
+}
+
+// GetState returns the encoded iterator state
+func (i *historyIterator) GetState() ([]byte, error) {
+	return json.Marshal(i.historyIteratorState)
+}
+
+func (i *historyIterator) readHistoryBatches(firstEventID int64) ([]*shared.History, historyIteratorState, error) {
+	size := 0
+	targetSize := i.config.TargetHistoryBatchesSize(i.domainName)
+	var historyBatches []*shared.History
+	newIterState := historyIteratorState{}
+	for size < targetSize {
+		currHistoryBatches, err := i.readHistory(firstEventID)
+		if _, ok := err.(*shared.EntityNotExistsError); ok && firstEventID != common.FirstEventID {
+			newIterState.FinishedIteration = true
+			return historyBatches, newIterState, nil
+		}
+		if err != nil {
+			return nil, newIterState, err
+		}
+		for idx, batch := range currHistoryBatches {
+			historyBatchSize, err := i.sizeEstimator.EstimateSize(batch)
+			if err != nil {
+				return nil, newIterState, err
+			}
+			size += historyBatchSize
+			historyBatches = append(historyBatches, batch)
+			firstEventID = *batch.Events[len(batch.Events)-1].EventId + 1
+
+			// If targetSize is meeted after appending the last batch, we are not sure if there's more batches or not,
+			// so we need to exclude that case.
```

Comment:
Can we rewrite this comment to be "In case targetSize is satisfied before reaching the end of current set of batches, return immediately"

Reply:
Updated. Here I'd like to emphasize why I excluded the last idx in the if statement.